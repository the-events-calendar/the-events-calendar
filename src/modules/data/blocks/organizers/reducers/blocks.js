/**
 * External dependencies
 */
import { combineReducers } from 'redux';

/**
 * Internal dependencies
 */
import * as types from './../types';

export const byId = ( state = {}, action ) => {
	switch ( action.type ) {
		case types.ADD_BLOCK_ORGANIZER:
			return {
				...state,
				[ action.payload.id ]: action.payload.organizer,
			};
		case types.REMOVE_BLOCK_ORGANIZER:
			return Object.keys( state ).reduce( ( newState, id ) => {
				if ( id === action.payload.id ) {
					return newState;
				}

				return {
					...newState,
					[ id ]: state[ id ],
				};
			}, {} );
		default:
			return state;
	}
};

export const allIds = ( state = [], action ) => {
	switch ( action.type ) {
		case types.ADD_BLOCK_ORGANIZER:
			return [ ...state, action.payload.organizer ];
		case types.REMOVE_BLOCK_ORGANIZER:
			return state.filter( ( organizer ) => organizer !== action.payload.organizer );
		default:
			return state;
	}
};

export default combineReducers( {
	byId,
	allIds,
} );
