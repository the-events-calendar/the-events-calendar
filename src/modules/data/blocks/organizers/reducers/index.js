export {
	default as classic,
	setInitialState as classicSetInitialState,
} from './classic';
export { default as blocks } from './blocks';
