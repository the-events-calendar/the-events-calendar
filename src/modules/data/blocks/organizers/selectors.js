/**
 * External dependencies
 */
import { createSelector } from 'reselect';
import { difference } from 'lodash';

export const getOrganizersInClassic = ( state ) => state.events.blocks.organizers.classic;

export const getOrganizerByClientId = ( state, props ) =>
	state.events.blocks.organizers.blocks.byId[ props.clientId ];

export const getOrganizersInBlock = ( state ) => state.events.blocks.organizers.blocks.allIds;

export const getMappedOrganizers = createSelector(
	[ getOrganizersInClassic, getOrganizersInBlock ],
	( classic, blocks ) => {
		return classic.map( ( id ) => ( {
			id,
			block: difference( [ id ], blocks ).length === 0,
		} ) );
	},
);
