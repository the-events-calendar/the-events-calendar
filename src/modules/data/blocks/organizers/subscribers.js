/**
 * External dependencies
 */
import { differenceBy } from 'lodash';

/**
 * Internal dependencies
 */
import { globals } from "@moderntribe/common/utils";
import { editor } from '@moderntribe/common/data';
import { store } from '@moderntribe/common/store';
import { actions as formActions } from '@moderntribe/common/data/forms';
import {
	actions as organizerActions,
	selectors as organizerSelectors,
} from '@moderntribe/events/data/blocks/organizers';
import { selectors as detailSelectors } from '@moderntribe/events/data/details';

const { getState, dispatch } = store;

/**
 * Returns criteria for comparing blocks.
 *
 * @exports
 * @param {Object} block Object with block attributes and data.
 *
 * @returns {string} Client ID of the block.
 */
export const compareBlocks = block => block.clientId;

/**
 * Checks whether the block is the organizer block.
 *
 * @exports
 * @param {Object} block Object with block attributes and data.
 *
 * @returns {boolean} Whether the block is the organizer block or not.
 */
export const isOrganizerBlock = ( block ) => block.name === 'tribe/event-organizer';

/**
 * Handles the block that was added.
 *
 * @exports
 * @param {Object} block Object with block attributes and data.
 */
export const handleBlockAdded = ( block ) => {
	// only handle event organizer block addition
	if ( ! isOrganizerBlock( block ) ) {
		return;
	}

	// return early if organizer attribute is empty
	if ( ! block.attributes.organizer ) {
		return;
	}

	dispatch( organizerActions.addOrganizerInBlock( block.clientId, block.attributes.organizer ) );
};

/**
 * Handles the block that was removed.
 *
 * @exports
 * @param {array} currBlocks Array of current blocks in the editor.
 *
 * @returns {Function} Function that handles the block that was removed.
 */
export const handleBlockRemoved = ( currBlocks ) => ( block ) => {
	// only handle event organizer block removal
	if ( ! isOrganizerBlock( block ) ) {
		return;
	}

	const classicBlock = currBlocks.filter( currBlock => currBlock.name === 'tribe/classic-event-details' );
	const organizer = organizerSelectors.getOrganizerByClientId( getState(), block );

	// remove organizer from block state
	if ( organizer ) {
		dispatch( organizerActions.removeOrganizerInBlock( block.clientId, organizer ) );
	}

	const volatile = detailSelectors.getVolatile( getState(), { name: organizer } );

	if ( ! classicBlock.length || volatile ) {
		// remove organizer from classic state
		dispatch( organizerActions.removeOrganizerInClassic( organizer ) );
		dispatch( formActions.removeVolatile( organizer ) );

		// set event organizer meta
		const classicOrganizers = organizerSelectors.getOrganizersInClassic( getState() );
		const postId = globals.wpData.select( 'core/editor' ).getCurrentPostId();
		const record = {
			meta: {
				_EventOrganizerID: classicOrganizers,
			},
		};

		globals.wpData.dispatch( 'core' ).editEntityRecord( 'postType', editor.EVENT, postId, record );
	}
};

/**
 * Handles changes in the blocks in the editor.
 *
 * @exports
 * @param {Array} currBlocks Array of current blocks in the editor.
 * @param {Array} prevBlocks Array of previous blocks in the editor.
 */
export const onBlocksChangeHandler = ( currBlocks, prevBlocks ) => {
	const blocksAdded = differenceBy( currBlocks, prevBlocks, compareBlocks );
	const blocksRemoved = differenceBy( prevBlocks, currBlocks, compareBlocks );

	if ( blocksAdded.length ) {
		blocksAdded.forEach( handleBlockAdded );
	}

	if ( blocksRemoved.length ) {
		blocksRemoved.forEach( handleBlockRemoved( currBlocks ) );
	}
};

/**
 * Listener for blocks change in the editor.
 *
 * @exports
 * @param {Function} selector Selector function to get current blocks.
 *
 * @returns {Function} Listener that subscribes to WP store.
 */
export const onBlocksChangeListener = ( selector ) => {
	let holdBlocks = selector();
	return () => {
		const prevBlocks = holdBlocks;
		const currBlocks = selector();
		holdBlocks = currBlocks;

		if (
			prevBlocks.length !== currBlocks.length ||
			differenceBy( currBlocks, prevBlocks, compareBlocks ).length
		) {
			onBlocksChangeHandler( currBlocks, prevBlocks )
		}
	};
};

/**
 * @function subscribe
 * @description This subscribes to any changes in the wp data store.
 *              Since state and attribute changes should not be called
 *              in `componentDidMount()` and `componentWillUnmount()`,
 *              global state changes (tribe common store) and meta changes
 *              are handled in a global subscriber.
 */
const subscribe = () => {
	globals.wpData.subscribe(
		onBlocksChangeListener(
			globals.wpDataSelectCoreEditor().getBlocks
		)
	);
};

export default subscribe;
