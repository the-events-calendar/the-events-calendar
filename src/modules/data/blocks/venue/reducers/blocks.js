/**
 * External dependencies
 */
import { combineReducers } from 'redux';
import { uniq } from 'lodash';

/**
 * Internal dependencies
 */
import * as types from './../types';

/**
 * Reducer that handles adding and removing blocks from state.
 *
 * This sets the state.events.blocks.venue.blocks.byId state, which holds all venue blocks indexed by clientId.
 *
 * @since 6.2.0
 * @param {Object} state State object.
 * @param {string} action Action being taken.
 * @returns {{}} The new state.
 */
export const byId = ( state = {}, action ) => {
	switch ( action.type ) {
		case types.ADD_BLOCK_VENUE:
			return {
				...state,
				[ action.payload.id ]: action.payload.venue,
			};
		case types.REMOVE_BLOCK_VENUE:
			return Object.keys( state ).reduce( ( newState, id ) => {
				if ( id === action.payload.id ) {
					return newState;
				}

				return {
					...newState,
					[ id ]: state[ id ],
				};
			}, {} );
		default:
			return state;
	}
};

/**
 * Reducer that handles adding and removing blocks from state.
 *
 * This sets the state.events.blocks.venue.blocks.allIds state, which is a numerically indexed array of all venues.
 *
 * @since 6.2.0
 * @param {Object} state State object.
 * @param {string} action Action being taken.
 * @returns {{}} The new state.
 */
export const allIds = ( state = [], action ) => {
	switch ( action.type ) {
		case types.ADD_BLOCK_VENUE:
			return uniq( [ ...state, action.payload.venue ] );
		case types.REMOVE_BLOCK_VENUE:
			return state.filter( venue => venue !== action.payload.venue );
		default:
			return state;
	}
};

/**
 * Reducer that handles setting the showMap and showMapLink values in state.
 *
 * This sets the state.events.blocks.venue.blocks.core state.
 *
 * @since 6.2.0
 * @param {Object} state State object.
 * @param {string} action Action being taken.
 * @returns {{}} The new state.
 */
export const core = ( state = {}, action ) => {
	switch ( action.type ) {
		case types.SET_VENUE_MAP:
			return {
				...state,
				showMap: action.payload.showMap,
			};
		case types.SET_VENUE_MAP_LINK:
			return {
				...state,
				showMapLink: action.payload.showMapLink,
			};
		default:
			return state;
	}
};

export default combineReducers( {
	core,
	byId,
	allIds,
} );
