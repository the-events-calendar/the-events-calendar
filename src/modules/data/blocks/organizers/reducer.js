/**
 * External dependencies
 */
import { combineReducers } from 'redux';

/**
 * Internal dependencies
 */
import { classic, classicSetInitialState, blocks } from './reducers';

export const setInitialState = ( data ) => {
	classicSetInitialState( data );
};

export default combineReducers( {
	blocks,
	classic,
} );
