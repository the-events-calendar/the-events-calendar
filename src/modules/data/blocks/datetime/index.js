/**
 * Internal dependencies
 */
import * as reducer from './reducer';
import * as types from './types';
import * as actions from './actions';
import * as selectors from './selectors';
import sagas from './sagas';

export default reducer.default;
export {
	reducer,
	types,
	actions,
	selectors,
	sagas,
};
