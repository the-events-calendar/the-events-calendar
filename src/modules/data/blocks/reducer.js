/**
 * External dependencies
 */
import { combineReducers } from 'redux';

/**
 * Internal dependencies
 */
import datetime, { reducer as datetimeReducer } from './datetime';
import organizers, { reducer as organizersReducer } from './organizers';
import price, { reducer as priceReducer } from './price';
import website, { reducer as websiteReducer } from './website';
import venue, { reducer as venueReducer } from './venue';

export const setInitialState = ( data ) => {
	datetimeReducer.setInitialState( data );
	organizersReducer.setInitialState( data );
	priceReducer.setInitialState( data );
	websiteReducer.setInitialState( data );
	venueReducer.setInitialState( data );
};

export default combineReducers( {
	datetime,
	venue,
	organizers,
	price,
	website,
} );
