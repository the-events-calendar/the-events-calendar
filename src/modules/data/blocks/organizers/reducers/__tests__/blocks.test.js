/**
 * Internal dependencies
 */
import blocks, { allIds, byId } from '@moderntribe/events/data/blocks/organizers/reducers/blocks';
import { actions } from '@moderntribe/events/data/blocks/organizers';

describe( '[STORE] - Organizer allIDs reducer', () => {
	test( 'Should return the default state', () => {
		expect( allIds( undefined, {} ) ).toMatchSnapshot();
	} );

	test( 'Should add organizer block', () => {
		expect( allIds( [], actions.addOrganizerInBlock( 99, 1 ) ) ).toMatchSnapshot();
		expect( allIds( [ 1 ], actions.addOrganizerInBlock( 100, 2 ) ) ).toMatchSnapshot();
	} );

	test( 'Should remove organizer block', () => {
		expect( allIds( [], actions.removeOrganizerInBlock( 102, 3 ) ) ).toMatchSnapshot();
		expect( allIds( [ 1, 2 ], actions.removeOrganizerInBlock( 100, 2 ) ) ).toMatchSnapshot();
		expect( allIds( [ 1 ], actions.removeOrganizerInBlock( 99, 1 ) ) ).toMatchSnapshot();
	} );
} );

describe( '[STORE] - Organizer byId reducer', () => {
	test( 'Should return the default state', () => {
		expect( byId( undefined, {} ) ).toEqual( {} );
	} );

	it( 'Should add organizer block', () => {
		expect( byId( {}, actions.addOrganizerInBlock( 99, 1 ) ) ).toMatchSnapshot();
		expect( byId( { 99: 1 }, actions.addOrganizerInBlock( 100, 2 ) ) ).toMatchSnapshot();
	} );

	it( 'Should remove organizer block', () => {
		expect( byId( {}, actions.removeOrganizerInBlock( 102, 3 ) ) ).toMatchSnapshot();
		expect( byId( { 99: 1, 100: 2 }, actions.removeOrganizerInBlock( 100, 2 ) ) ).toMatchSnapshot();
		expect( byId( { 99: 1 }, actions.removeOrganizerInBlock( 99, 1 ) ) ).toMatchSnapshot();
	} );
} );

describe( '[STORE] - Organizer blocks reducer', () => {
	test( 'Should return the default state', () => {
		expect( blocks( undefined, {} ) ).toMatchSnapshot();
	} );
} );
