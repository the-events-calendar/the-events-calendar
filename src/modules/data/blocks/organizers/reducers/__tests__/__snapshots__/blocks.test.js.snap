// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[STORE] - Organizer allIDs reducer Should add organizer block 1`] = `
[
  1,
]
`;

exports[`[STORE] - Organizer allIDs reducer Should add organizer block 2`] = `
[
  1,
  2,
]
`;

exports[`[STORE] - Organizer allIDs reducer Should remove organizer block 1`] = `[]`;

exports[`[STORE] - Organizer allIDs reducer Should remove organizer block 2`] = `
[
  1,
]
`;

exports[`[STORE] - Organizer allIDs reducer Should remove organizer block 3`] = `[]`;

exports[`[STORE] - Organizer allIDs reducer Should return the default state 1`] = `[]`;

exports[`[STORE] - Organizer blocks reducer Should return the default state 1`] = `
{
  "allIds": [],
  "byId": {},
}
`;

exports[`[STORE] - Organizer byId reducer Should add organizer block 1`] = `
{
  "99": 1,
}
`;

exports[`[STORE] - Organizer byId reducer Should add organizer block 2`] = `
{
  "100": 2,
  "99": 1,
}
`;

exports[`[STORE] - Organizer byId reducer Should remove organizer block 1`] = `{}`;

exports[`[STORE] - Organizer byId reducer Should remove organizer block 2`] = `
{
  "100": 2,
  "99": 1,
}
`;

exports[`[STORE] - Organizer byId reducer Should remove organizer block 3`] = `
{
  "99": 1,
}
`;
